doc/languages-frameworks/maven.md

```diff
@@ -0,0 +1,298 @@
+          == "target");
+      src = ./.; # or fetchFromGitHub, cleanSourceWith, etc
+    };
+  };
+  buildInputs = [ maven ];
+
+  buildPhase = ''
+    echo "Using repository ${repository}"
+    mvn --offline -Dmaven.repo.local=${repository} package;
+  '';
+
+  installPhase = ''
+    mkdir $out/share/java
+    cp target/${pname}-${version}.jar $out/share/java
+  '';
+}
+```
+
+> We place the library in _$out/share/java_ since JDK have a _stdenv setup hook_ that add any JARs in the _share/java_ directories of the build inputs to the CLASSPATH environment.
+
+## Runnable JAR
+
+You are able to run `mvn compile` but how do you get a runnable artifact without having to use something like building an UberJar ?
+
+We will use the same repository we built above (either _double invocation_ or _buildMaven_) to setup a CLASSPATH for our JAR.
+
+Augment the _pom.xml_ to create a JAR with the following manifest:
+```xml
+<build>
+  <plugins>
+    <plugin>
+        <artifactId>maven-jar-plugin</artifactId>
+        <configuration>
+            <archive>
+                <manifest>
+                    <addClasspath>true</addClasspath>
+                    <classpathPrefix>lib/</classpathPrefix>
+                    <classpathLayoutType>repository</classpathLayoutType>
+                    <mainClass>Main</mainClass>
+                </manifest>
+                <manifestEntries>
+                    <Class-Path>.</Class-Path>
+                </manifestEntries>
+            </archive>
+        </configuration>
+    </plugin>
+  </plugins>
+</build>
+```
+
+> This instructs the JAR to look for the necessary dependencies in the _lib/_ relative folder. The layout of the folder is also in the _maven repository_ style.
+
+We will modify the derivation above to add a symlink to our repository so that it's accessible to our JAR during the _installPhase_.
+
+```nix
+{ pkgs ? import <nixpkgs> { }, stdenv ? pkgs.stdenv, lib ? pkgs.lib
+, maven ? pkgs.maven, jre ? pkgs.jre, makeWrapper ? pkgs.makeWrapper }:
+# pick a repository derivation from above
+let repository = ...;
+in stdenv.mkDerivation rec {
+  pname = "maven-demo";
+  version = "1.0";
+
+  src = lib.cleanSourceWith {
+    filter = lib.cleanSourceFilter;
+    src = lib.cleanSourceWith {
+      filter = path: type:
+        !(lib.pathIsDirectory (path) && baseNameOf (toString (path))
+          == "target");
+      src = ./.; # or fetchFromGitHub, cleanSourceWith, etc
+    };
+  };
+  buildInputs = [ maven makeWrapper];
+
+  buildPhase = ''
+    echo "Using repository ${repository}"
+    mvn --offline -Dmaven.repo.local=${repository} package;
+  '';
+
+  installPhase = ''
+    # create the bin directory
+    mkdir -p $out/bin
+    # create a symbolic link for the lib directory
+    ln -s ${repository} $out/lib
+    # copy out the JAR
+    # Maven already setup the classpath to use m2 repository layout
+    # with the prefix of lib/
+    cp target/${pname}-${version}.jar $out/
```

Comment:
Shouldn't this be `$out/share/java/` as in "Building a JAR"?

Reply:
Fixed